### PR TITLE
Turn NodeId into a newtype, refactor findTopShape

### LINKED ISSUE #1002: NodeId should be a newtype
https://github.com/google/codeworld/blob/24778237c19a39e77b415d43dc8e681493a0e539/codeworld-api/src/CodeWorld/Driver.hs#L295-L298

Defining `NodeId` as a type alias makes the code harder to follow. When `Int`s appear in type signatures it is not clear whether they refer to `NodeId`s or not.

### PR DESCRIPTION
**Pull request by sjakobi**
Fixes #1002.

### CODE REVIEW

File: codeworld-api/src/CodeWorld/Driver.hs [hunk 1]
```diff
@@ -659,20 +661,26 @@ drawDrawing ds (Shape shape) = drawShape $ shape ds
...
-        False -> return (False, 1)
-findTopShape ds (Transformation f d) =
-    fmap (+ 1) <$> findTopShape (f ds) d
-findTopShape ds (Drawings []) = return (False, 1)
-findTopShape ds (Drawings (dr:drs)) = do
-    (found, count) <- findTopShape ds dr
-    case found of
-        True -> return (True, count + 1)
-        False -> fmap (+ count) <$> findTopShape ds (Drawings drs)
+findTopShape :: MonadCanvas m => DrawState -> Drawing m -> m (Maybe NodeId)
```
Comment (sjakobi) [inline codeworld-api/src/CodeWorld/Driver.hs]: It turned out that we never cared about the Int when the Bool was `False`.

File: codeworld-api/src/CodeWorld/Driver.hs [hunk 2]
```diff
@@ -1738,34 +1743,34 @@ highlightDrawing (DrawState at _) drawing =
...
-        case go ds (n - 1) dr of
+        case go ds (pred n) dr of
             Left d -> Left d
-            Right n -> go ds (n + 1) $ Drawings drs
+            Right n -> go ds (succ n) $ Drawings drs
 
 replaceDrawNode :: forall m. NodeId -> Drawing m -> Drawing m -> Maybe (Drawing m)
 replaceDrawNode n _ _
-    | n < 0 = Nothing
+    | n < (NodeId 0) = Nothing
```
Comment (sjakobi) [inline codeworld-api/src/CodeWorld/Driver.hs]: I'm not quite sure why we have this case checking for an invalid NodeId. Maybe we should use `error` here?

File: codeworld-api/src/CodeWorld/Driver.hs [hunk 3]
```diff
@@ -922,7 +927,7 @@ initDebugMode setActive getPic highlight = do
             -- because handlePointRequest ignores them and works only with
             -- an off-screen image.
             n <- runCanvasM undefined undefined (handlePointRequest getPic (x, y))
-            return (pToJSVal (fromMaybe (-1) n))
+            return (pToJSVal (maybe (-1) getNodeId n))
```
Comment (sjakobi) [inline codeworld-api/src/CodeWorld/Driver.hs]: Signaling with an invalid NodeId is a bit ugly… :/

**Comment by sjakobi** [comment]
I just re-checked that the the inspection dialog still works!

So this should be good to go unless you have any suggestions on my comments above or any other concerns.

File: codeworld-api/src/CodeWorld/Driver.hs (hunk 2 above)
Comment (cdsmith) [reply codeworld-api/src/CodeWorld/Driver.hs]: I don't know.  I think the front end code uses -1 to indicate no node is selected, so this might be used.

**Pull request merged**